component/wrapper.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2016-2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package component
+
+import (
+	p "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/infer"
+	"github.com/pulumi/pulumi-go-provider/middleware/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+type registry struct {
+	// inferredComponents is a set of inferred components
+	inferredComponents map[Resource]struct{}
+}
+
+// globalRegistry is a global registry for all inferred components. This is used to
+// register the components with the provider during initialization.
+var globalRegistry = registry{
+	inferredComponents: make(map[Resource]struct{}),
+}
+
+// Resource is a type alias for the inferred component resource type.
+type Resource = infer.InferredComponent
+
+// ConstructorFn is a type alias for infer.ConstructorFn which represents
+// a function that creates a component resource.
+type ConstructorFn[I any, O pulumi.ComponentResource] = infer.ComponentFn[I, O]
+
+// ProgramComponent allows us to create a component resource that can be understood by the
+// underlying provider. This is a wrapper around the infer.ProgramComponent function.
+func ProgramComponent[I any, O pulumi.ComponentResource](fn ConstructorFn[I, O]) Resource {
+	return infer.ProgramComponent[I, O](fn)
+}
+
+// RegisterType registers a type within the global registry for this wrapper package. This
+// allows the provider to access the custom types and functions required to infer its schema.
+func RegisterType(ic Resource) {
```

Comment:
A helpful sanity check would be to panic if someone tries to call this after calling `ProviderHost`

Reply:
I'm not sure if this is necessary since `ProviderHost` starts a blocking gRPC server exposing the provider. If for some reason the ProviderHost is stopped, there may be a use case in which more types should be potentially added.

Reply:
That's fair, though people can do surprising things. (someone might try to registry a new component from inside the body of another component constructor?) I tend to prefer being rather noisy about the contract for these sorts of global context facilities, but you have a fair point that this is unlikely. 

(Another edge case might be to warn loudly somehow if there are no components registered in the list when invoking `ProviderHost`; eg. someone could try adding components in main after calling `ProviderHost`.

Reply:
That's fair. I've updated it so that there needs to be at least 1 inferred type before a provider starts, and that we can only start the provider once, along with preventing type registration after calling `ProviderHost`. Should there be an actual usecase for this, we can always relax these restrictions.